File: acmp/binsearch/G. Вырубка леса/woodcutting.py
def test(a, b, k, m, x, days):
    s = a * days - a * (days // k); 
    s += b * days - b * (days // m); 
    return s >= x

a, k, b, m, x = [int(x) for x in input().split()]

low = 0
high = 10**18

while low < high:
    days = (low + high) // 2
    if not test(a, b, k, m, x, days): 
        low = days + 1
    else:
        high = days

print(low)

File: acmp/binsearch/G. Вырубка леса/woodcutting.cpp
#include <iostream>

using namespace std;

using ull = uint64_t;

ull a, k, b, m, x;

bool test(ull days) {
    ull s = a * (days - (days / k));
    s += b * (days - (days / m));

    return s >= x;
}

int main() {
    cin >> a >> k >> b >> m >> x;

    ull low = 0, high = 1;
    //high = 1e18 может вызвать переполнение в a*high,
    //поэтому ищем адекватную верхнюю грань
    while (!test(high)) high *= 2;

    while (low < high) {
        ull mid = (low + high) / 2;
        if (!test(mid)) {
            low = mid + 1;
        }
        else {
            high = mid;
        }
    }

    cout << low << endl;
}

